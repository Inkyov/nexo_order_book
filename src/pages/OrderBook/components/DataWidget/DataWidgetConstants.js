import keyMirror from "key-mirror-nested";

export const STREAM_LABELS = {
  btcusdt: { price: "USDT", amount: "BTC" },
  ethbtc: { price: "BTC", amount: "ETH" },
};

export const COMPONENT_TEST_IDS = keyMirror({
  PRICE: null,
  AMOUNT: null,
  ORDER_ROW: null,
});
