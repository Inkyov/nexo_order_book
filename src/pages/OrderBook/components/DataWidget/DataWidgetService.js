export const formatPrice = ({ currentStream, value }) =>
  parseFloat(value).toFixed(currentStream === "btcusdt" ? 2 : 6);

export const formatAmount = ({ currentStream, value }) =>
  parseFloat(value).toFixed(currentStream === "btcusdt" ? 6 : 3);
