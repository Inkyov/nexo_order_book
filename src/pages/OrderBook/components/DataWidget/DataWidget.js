import React from "react";
import { isEmpty } from "lodash/fp";
import { Grid, styled } from "@material-ui/core";
import PropTypes from "prop-types";
import { formatAmount, formatPrice } from "./DataWidgetService";
import { STREAM_LABELS, COMPONENT_TEST_IDS } from "./DataWidgetConstants";

const GridItem = styled(Grid)(({ theme }) => ({
  flex: 1,
  marginBottom: theme.spacing(1),
}));

const DataWidget = ({ data, currentStream, ...props }) => {
  const labels = STREAM_LABELS[currentStream];

  return (
    <Grid item xs={6} {...props}>
      <Grid container>
        <GridItem item xs={6} data-testid={COMPONENT_TEST_IDS.PRICE}>
          Price({labels.price})
        </GridItem>
        <GridItem item xs={6} data-testid={COMPONENT_TEST_IDS.AMOUNT}>
          Amount({labels.amount})
        </GridItem>
      </Grid>
      {!isEmpty(data) &&
        data.map((entry, index) => (
          <Grid
            container
            key={index}
            data-testid={`${COMPONENT_TEST_IDS.ORDER_ROW}_${index}`}
          >
            <Grid item xs={6}>
              {formatPrice({ currentStream, value: entry[0] })}
            </Grid>
            <Grid item xs={6}>
              {formatAmount({ currentStream, value: entry[1] })}
            </Grid>
          </Grid>
        ))}
    </Grid>
  );
};

DataWidget.propTypes = {
  data: PropTypes.array,
  currentStream: PropTypes.string.isRequired,
};

export default DataWidget;
