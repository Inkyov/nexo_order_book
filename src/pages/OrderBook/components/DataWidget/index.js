import DataWidget from "./DataWidget";

export default DataWidget;
