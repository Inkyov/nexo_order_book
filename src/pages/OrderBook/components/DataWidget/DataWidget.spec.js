import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup, render } from "@testing-library/react";
import DataWidget from "./DataWidget";
import { COMPONENT_TEST_IDS } from "./DataWidgetConstants";

const defaultProps = {
  data: [
    ["0.33333", "0.12345"],
    ["0.222222", "0.12345"],
  ],
  currentStream: "btcusdt",
};

const renderComponent = (props) =>
  render(<DataWidget {...{ ...defaultProps, ...props }} />);

describe("DataWidget", () => {
  beforeEach(cleanup);

  it("should render", () => {
    const { container } = renderComponent();

    expect(container).toMatchSnapshot();
  });

  it("should render price label", () => {
    const { getByTestId } = renderComponent();
    const priceLabel = getByTestId(COMPONENT_TEST_IDS.PRICE);
    expect(priceLabel).toBeVisible();
  });

  it("should render amount label", () => {
    const { getByTestId } = renderComponent();
    const amountLabel = getByTestId(COMPONENT_TEST_IDS.AMOUNT);
    expect(amountLabel).toBeVisible();
  });

  it("should render rows", () => {
    const { getByTestId } = renderComponent();

    const orderRow = getByTestId(`${COMPONENT_TEST_IDS.ORDER_ROW}_0`);
    expect(orderRow).toBeVisible();
  });
});
