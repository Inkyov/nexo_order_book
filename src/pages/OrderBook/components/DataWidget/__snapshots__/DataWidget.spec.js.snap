// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataWidget should render 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
  >
    <div
      class="MuiGrid-root MuiGrid-container"
    >
      <div
        class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-1 WithStyles(ForwardRef(Grid))-root-2 MuiGrid-item MuiGrid-grid-xs-6"
        data-testid="PRICE"
      >
        Price(
        USDT
        )
      </div>
      <div
        class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-1 WithStyles(ForwardRef(Grid))-root-3 MuiGrid-item MuiGrid-grid-xs-6"
        data-testid="AMOUNT"
      >
        Amount(
        BTC
        )
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container"
      data-testid="ORDER_ROW_0"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      >
        0.33
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      >
        0.123450
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container"
      data-testid="ORDER_ROW_1"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      >
        0.22
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      >
        0.123450
      </div>
    </div>
  </div>
</div>
`;
